File: client-java/instrumentation/src/main/java/org/evomaster/clientJava/instrumentation/ClassName.java
package org.evomaster.clientJava.instrumentation;

import java.util.Objects;

public class ClassName {

    /**
     * Name used in the bytecode, where "/" are used
     * instead of "."
     * eg
     * org.bar.Foo turns into org/bar/Foo
     */
    private final String bytecodeName;

    /**
     * What usually returned with Foo.class.getName(),
     * eg
     * org.bar.Foo
     */
    private final String fullNameWithDots;

    public static ClassName get(Class<?> klass){
        return new ClassName(klass);
    }

    public static ClassName get(String name) {
        return new ClassName(name);
    }

    public ClassName(Class<?> klass){
        this(Objects.requireNonNull(klass).getName());
    }

    /**
     *
     * @param name of the class, or path resource
     */
    public ClassName(String name){
        Objects.requireNonNull(name);

        if(name.endsWith(".class")){
            name = name.substring(0, name.length() - ".class".length());
        }
        if(name.endsWith(".java")){
            name = name.substring(0, name.length() - ".java".length());
        }

        if(name.contains("/") && name.contains(".")){
            throw new IllegalArgumentException("Do not know how to handle name: " +name);
        }

        if(name.contains("/")){
            bytecodeName = name;
            fullNameWithDots = name.replace("/", ".");
        } else {
            bytecodeName = name.replace(".", "/");
            fullNameWithDots = name;
        }
    }

    /**
        Name of the class as used in the bytecode instructions.
        This means that foo.bar.Hello would be foo/bar/Hello
     */
    public String getBytecodeName() {
        return bytecodeName;
    }

    /**
     * Eg, foo.bar.Hello
     * @return
     */
    public String getFullNameWithDots() {
        return fullNameWithDots;
    }

    public String getAsResourcePath(){
        return bytecodeName + ".class";
    }

    @Override
    public String toString(){
        return "[" + this.getClass().getSimpleName()+": "+fullNameWithDots+"]";
    }
}

File: client-dotnet/src/EvoMaster.Instrumentation_Shared/ClassName.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using EvoMaster.Client.Util.Extensions;

namespace EvoMaster.Instrumentation_Shared {
    public class ClassName {
        /*
        Name used in the bytecode, where "/" are used
        instead of "."
        eg
        org.bar.Foo turns into org/bar/Foo
        */
        private readonly string _bytecodeName;

        /*
        What usually returned with Foo.class.getName(),
        eg
        org.bar.Foo
        */
        private readonly string _fullNameWithDots;


        private static readonly IDictionary<Type, ClassName>
            CacheClass = new ConcurrentDictionary<Type, ClassName>(); //TODO: capacity 10_000

        private static readonly IDictionary<string, ClassName> Cachestring =
            new ConcurrentDictionary<string, ClassName>(); //TODO: capacity 10_000

        public static ClassName Get(Type klass) {
            return CacheClass.ComputeIfAbsent(klass, k => new ClassName(k));
            //return new ClassName(klass);
        }

        public static ClassName Get(string name) {
            return Cachestring.ComputeIfAbsent(name, n => new ClassName(n));
            //return new ClassName(name);
        }

        public ClassName(Type klass) : this(klass.RequireNonNull<Type>().Name) { }

        /// <param name="name">name of the class, or path resource</param>
        public ClassName(string name) {
            name.RequireNonNull<string>();

            if (name.EndsWith(".class")) {
                name = name.Substring(0, name.Length - ".class".Length);
            }

            if (name.EndsWith(".cs")) {
                name = name.Substring(0, name.Length - ".cs".Length);
            }

            if (name.Contains("/") && name.Contains(".")) {
                throw new ArgumentException("Do not know how to handle name: " + name);
            }

            if (name.Contains("/")) {
                _bytecodeName = name;
                _fullNameWithDots = name.Replace("/", ".");
            }
            else {
                _bytecodeName = name.Replace(".", "/");
                _fullNameWithDots = name;
            }
        }

        ///<summary>Name of the class as used in the bytecode instructions.
        /// This means that foo.bar.Hello would be foo/bar/Hello</summary>
        public string GetBytecodeName() {
            return _bytecodeName;
        }

        /// <summary>Eg, foo.bar.Hello</summary>
        public string GetFullNameWithDots() {
            return _fullNameWithDots;
        }

        public string GetAsResourcePath() {
            return _bytecodeName + ".class";
        }


        public override string ToString() {
            return "[" + this.GetType().Name + ": " + _fullNameWithDots + "]";
        }
    }
}